File: 2017/test/day_02_test.exs
defmodule AdventOfCode2017.Day02Test do
  use ExUnit.Case

  import AdventOfCode2017.Day02
  import AdventOfCode2017.TestHelper

  describe "Part 1" do
    test "part1" do
      assert part1(input_sample_1()) == 18
      assert part1(input_sample_2()) == 72
      assert part1(input("test/input/day_02.txt")) == 45158
    end
  
    defp input_sample_1, do: "5 1 9 5\n7 5 3\n2 4 6 8"
    defp input_sample_2, do: "12 8 4 1\n1 62 5"
  end

  describe "Part 2" do
    test "part2" do
      assert part2(input_sample_1_2()) == 9
      assert part2(input("test/input/day_02.txt")) == 294
    end

    test "get_row_divided" do
      assert get_row_divided([5, 9, 2, 8]) == 4
    end

    defp input_sample_1_2, do: "5 9 2 8\n9 4 7 3\n3 8 6 5"
  end
end

File: 2018/test/day_02_test.rb
require "test_helper"
require "day_02"

class Day02Test < Minitest::Test
  def setup
    @day_02 = Day02.new
  end

  def test_has_count?
    assert_equal false, @day_02.has_count?("abcdef", 2)
    assert_equal true, @day_02.has_count?("bababc", 2)
    assert_equal true, @day_02.has_count?("abbcde", 2)
    assert_equal false, @day_02.has_count?("abcccd", 2)
    assert_equal true, @day_02.has_count?("aabcdd", 2)
    assert_equal true, @day_02.has_count?("abcdee", 2)
    assert_equal false, @day_02.has_count?("ababab", 2)

    assert_equal false, @day_02.has_count?("abcdef", 3)
    assert_equal true, @day_02.has_count?("bababc", 3)
    assert_equal false, @day_02.has_count?("abbcde", 3)
    assert_equal true, @day_02.has_count?("abcccd", 3)
    assert_equal false, @day_02.has_count?("aabcdd", 3)
    assert_equal false, @day_02.has_count?("abcdee", 3)
    assert_equal true, @day_02.has_count?("ababab", 3)
  end

  def test_two_count
    assert_equal 4, @day_02.num_count(sample_input_1.split, 2)
  end

  def test_three_count
    assert_equal 3, @day_02.num_count(sample_input_1.split, 3)
  end

  def test_part1
    assert_equal 12, @day_02.part1(sample_input_1)
    assert_equal 6370, @day_02.part1(input("day_02.txt"))
  end

  def test_difference_count
    assert_equal 2, @day_02.difference_count("abcde", "axcye")
    assert_equal 1, @day_02.difference_count("fghij", "fguij")
  end

  def test_remove_differing_characters
    assert_equal "ace", @day_02.remove_differing_characters("abcde", "axcye")
    assert_equal "fgij", @day_02.remove_differing_characters("fghij", "fguij")
  end

  def test_part2
    assert_equal "fgij", @day_02.part2(sample_input_2)
    # assert_equal "rmyxgdlihczskunpfijqcebtv", @day_02.part2(input("day_02.txt"))
  end

  def sample_input_1
    "abcdef
    bababc
    abbcde
    abcccd
    aabcdd
    abcdee
    ababab"
  end

  def sample_input_2
    "abcde
    fghij
    klmno
    pqrst
    fguij
    axcye
    wvxyz"
  end
end
